subsys/instrumentation/Kconfig

```diff
@@ -0,0 +1,96 @@
+# Copyright 2023 Linaro
+# SPDX-License-Identifier: Apache-2.0
+
+# Instrumentation configuration options
+
+config INSTRUMENTATION
+	bool "Compiler Instrumentation Support"
+	imply REBOOT
+	imply RETAINED_MEM
+	imply RETAINED_MEM_ZEPHYR_RAM
+	imply RETENTION
+	imply THREAD_MONITOR
+	imply THREAD_NAME
+	imply THREAD_STACK_INFO
+	imply TIMING_FUNCTIONS
+	imply TRACING
+	imply UART_INTERRUPT_DRIVEN
```

Comment:
I've cleaned it up a bit. `THREAD_MONITOR` and `THREAD_STACK_INFO` are not necessary indeed. Also, I fixed the `implies` since they should actually be `selects`.

Based on what we've discussed in the ZDS, I've split the tracing and profiling parts since the "flame graphs" can be deduced from the traces alone when a small region is traced (I still think that getting statistical profile data as in perf is valuable as I tried to explain). Now the tracing part (`INSTRUMENTATION_MODE_CALLGRAPH`) can be enabled/disabled independently from the tracing part (`INSTRUMENTATION_MODE_STATISTICAL`).

The new Kconfig is more organized now, I believe.

So, the justification for the remaining selected CONFIGs is:

To be able to use the CLI tool to set the trigger and stopper addresses at runtime it's necessary the retention subsys because once these addresses are set the target needs to be rebooted (hence `REBOOT`) and the addresses can not be overwritten, hence all the `RETAINED_*`  and `RETENTION_*` CONFIGs.

`TRACING` is to make `z_thread_mark_switched_in` and `z_thread_mark_switched_out` functions available, for detecting the context switches. Also, possibly in the future, for allowing probing of other hooks when tracing.